### PR TITLE
OpenPaaS-Suite/esn-frontend-calendar#473 bootstrap the ICS object to Event Object parser

### LINKED ISSUE #473: [TASK] Create a ICS object to Event Object parser
Create ICS to Event Object converter

Event Object should  [FullCalendar Event Object](https://fullcalendar.io/docs/event-object)

QUESTIONS:

Do we leverage is Ical.js ?

### PR DESCRIPTION
**Pull request by rezk2ll**
resolves https://github.com/OpenPaaS-Suite/esn-frontend-calendar/issues/473

### CODE REVIEW

**Comment by itswillta** [comment]
Why don't you call it `dav-parser `so that we can use this same repo for contacts stuff also.

**Comment by itswillta** [comment]
You forgot to link the issue.

File: package.json [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    "lint": "eslint src --ext .ts 'src/**/*.ts' 'test/**/*.ts'",
+    "lint:fix": "npm run lint -- --fix",
+    "prebuild": "rimraf dist",
+    "build": "tsc --module commonjs && rollup -c rollup.config.ts",
+    "start": "rollup -c rollup.config.ts -w",
+    "test": "jest --coverage",
+    "test:watch": "jest --coverage --watch",
+    "test:prod": "npm run lint && npm run test -- --no-cache"
+  },
+  "jest": {
```
Comment (itswillta) [inline package.json]: I think we should have the jest configuration in a separate file.

File: rollup.config.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+import typescript from '@rollup/plugin-typescript'
+import { terser } from 'rollup-plugin-terser'
+
+const pkg = require('./package.json')
+
+export default {
+  input: `src/index.ts`,
+  output: [
+    { file: pkg.main, name: pkg.name , format: 'umd', sourcemap: true },
+    { file: pkg.module, format: 'es', sourcemap: true, plugins: [terser()] },
```
Comment (itswillta) [inline rollup.config.ts]: Why only terser for ES modules?

File: src/lib/parser.ts [hunk 3]
```diff
@@ -0,0 +1,25 @@
+import * as ICAL from 'ical.js';
+import { FullCalendarEventObject } from '../types/caldav-parser';
+
+export function parse (ics: string): FullCalendarEventObject[] {
```
Comment (itswillta) [inline src/lib/parser.ts]:
```suggestion
export function parse(ics: string): FullCalendarEventObject[] {
```

File: src/types/caldav-parser.d.ts [hunk 4]
```diff
@@ -0,0 +1,62 @@
+
```
Comment (itswillta) [inline src/types/caldav-parser.d.ts]:
```suggestion
```

File: package.json [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    "nodemon": "^2.0.7",
+    "rimraf": "^2.6.2",
+    "rollup": "2.42.0",
+    "rollup-plugin-sourcemaps": "^0.4.2",
+    "rollup-plugin-terser": "^7.0.2",
+    "ts-jest": "^26.5.4",
+    "typescript": "^4.2.0"
+  },
+  "dependencies": {
+    "@types/rrule": "^2.2.9",
```
Comment (itswillta) [inline package.json]: This should be in `devDependencies`.

And where's the actual `rrule` package? I don't see it in `package.json`.

File: package.json [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+  "devDependencies": {
+    "@rollup/plugin-commonjs": "^17.1.0",
+    "@rollup/plugin-node-resolve": "^11.2.0",
+    "@rollup/plugin-typescript": "^8.2.0",
+    "@types/jest": "^23.3.2",
+    "@typescript-eslint/eslint-plugin": "^4.19.0",
+    "eslint": "^7.22.0",
+    "jest": "^26.6.3",
+    "jest-config": "^26.6.3",
+    "nodemon": "^2.0.7",
```
Comment (itswillta) [inline package.json]: Why `nodemon`?

File: package.json [hunk 7]
```diff
@@ -0,0 +1,78 @@
+{
+  "name": "caldav-parser",
+  "version": "0.0.0",
+  "description": "",
+  "keywords": [],
+  "main": "dist/caldav-parser.umd.js",
```
Comment (itswillta) [inline package.json]: I think we'll probably only use ESM in our future applications.

File: src/lib/parser.ts [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+
+  return subComponents.map(componentToEvent);
+}
+
+
+const componentToEvent = (component: ICAL.Component): FullCalendarEventObject => {
+  const event: ICAL.Event = new ICAL.Event(component);
+  const fullcalendarEvent: FullCalendarEventObject = {
+    id: event.uid,
+    allDay: false,
```
Comment (itswillta) [inline src/lib/parser.ts]: Shouldn't we have some comments about this?

Comment (rezk2ll) [reply src/lib/parser.ts]: ye not finished yet :laughing:

File: package.json (hunk 5 above)
Comment (rezk2ll) [reply package.json]: just needed the type

Comment (itswillta) [reply package.json]: OK... that's a little bit weird though :joy:.

File: .gitignore [hunk 9]
```diff
@@ -0,0 +1,12 @@
...
+.nyc_output
+.DS_Store
+*.log
+.vscode
+.idea
+dist
+compiled
+.awcache
+.rpt2_cache
+docs
```
Comment (itswillta) [inline .gitignore:12]: How about `package-lock.json`?

File: src/types/ical.js.d.ts [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+    constructor(options: {
+      /** Start time of the event */
+      dtstart: Time;
+      /** Component for expansion, required if not resuming. */
+      component?: Component;
+    })
+  }
+
+  export class Timezone {
+    public static  utcTimezone: Timezone;
```
Comment (itswillta) [inline src/types/ical.js.d.ts]:
```suggestion
    public static utcTimezone: Timezone;
```

File: package.json (hunk 6 above)
Comment (rezk2ll) [reply package.json]: oops, i was testing stuff with it and forgot to remove it

File: src/types/caldav-parser.d.ts [hunk 11]
```diff
@@ -0,0 +1,41 @@
+import { RRule } from 'rrule'
+
+export interface FullCalendarEventObject {
```
Comment (chamerling) [inline src/types/caldav-parser.d.ts]: Why do we have a `FulCalendar` object in a lib which is named `caldav-parser`?

Comment (rezk2ll) [reply src/types/caldav-parser.d.ts]: the goal is to provide a calendar object that is similar/acceptable by fullcalendar 5.x
https://fullcalendar.io/docs/event-parsing

and yes the name of the lib is somewhat not been decided yet

Comment (chamerling) [reply src/types/caldav-parser.d.ts]: OK yes, but do nit name it `FullCalendarEventObject`, name it `CalendarEventObject`

Comment (rezk2ll) [reply src/types/caldav-parser.d.ts]: sounds better, thank you

File: test/parser.test.ts [hunk 12]
```diff
@@ -0,0 +1,51 @@
+import { parse } from '../src/lib/parser';
+import { CalendarEventObject, Attendee } from '../src/types/dav-parser';
+import { simpleIcs, customPropertiesIcs, recurringEventIcs } from './const';
+
+describe('the parse function', () => {
+  const assertAttendeeParsing = (attendee: Attendee, partstat:string, email:string, cn:string) => {
```
Comment (itswillta) [inline test/parser.test.ts]:
```suggestion
  const assertAttendeeParsing = (attendee: Attendee, partstat: string, email: string, cn: string) => {
```

File: src/types/ical.js.d.ts [hunk 13]
```diff
@@ -0,0 +1,212 @@
...
+    public getFirstSubcomponent(name?: string): Component | null;
+    public getAllSubcomponents(name?: string): Component[];
+
+    public getFirstPropertyValue<T = any>(name?: string): T;
+
+    public getFirstProperty(name?: string): Property;
+    public getAllProperties(name?: string): Property[];
+
+    public addProperty(property: Property): Property;
+    public addPropertyWithValue(name: string,value: string | number | Record<string, unknown>): Property;
```
Comment (itswillta) [inline src/types/ical.js.d.ts]:
```suggestion
    public addPropertyWithValue(name: string, value: string | number | Record<string, unknown>): Property;
```

File: src/types/ical.js.d.ts [hunk 14]
```diff
@@ -0,0 +1,212 @@
...
+
+    public hasProperty(name?: string): boolean;
+
+    public updatePropertyWithValue(name: string,value: string | number | Record<string, unknown>): Property;
```
Comment (itswillta) [inline src/types/ical.js.d.ts]:
```suggestion
    public updatePropertyWithValue(name: string, value: string | number | Record<string, unknown>): Property;
```

File: README.md [hunk 15]
```diff
@@ -1 +1,9 @@
-# calendar ics parser
\ No newline at end of file
+# dav ics parser
+
+### usage
+
+```js
+import { parser } from 'dav-parser';
+
+const events = parse(ics);
```
Comment (itswillta) [inline README.md]:
```suggestion
import { parse } from 'dav-parser';

const events = parse(ics);
```

File: README.md [hunk 16]
```diff
@@ -1 +1,9 @@
-# calendar ics parser
\ No newline at end of file
+# dav ics parser
```
Comment (itswillta) [inline README.md]:
```suggestion
# DAV ics parser
```

File: README.md [hunk 17]
```diff
@@ -1 +1,9 @@
...
+
+### usage
```
Comment (itswillta) [inline README.md]:
```suggestion
## Usage
```

File: jest.config.ts [hunk 18]
```diff
@@ -0,0 +1,16 @@
+import type {Config} from '@jest/types';
+
```
Comment (itswillta) [inline jest.config.ts]:
```suggestion
import type { Config } from '@jest/types';
```

**Pull request merged**